### PR TITLE
Bug fix installs

### PR DESCRIPTION
**Pull request by KZhu-ME**
`installparticlesolver` seems to be working, tested using a dummy print function.
Implemented `installfieldsolver` callback based on `direct_solver` branch, but had to remove `#ifdef AMREX_USE_EB` statements in WarpXEvolve.cpp as the install didn't seem to work when the statements were there.

### CODE REVIEW

File: Source/Evolve/WarpXEvolve.cpp [hunk 1]
```diff
@@ -241,6 +241,7 @@ WarpX::Evolve (int numsteps)
             mypc->SortParticlesByBin(sort_bin_size);
         }
 
+        if (warpx_py_beforeEsolve) warpx_py_beforeEsolve();
```
Comment (peterscherpelz) [inline Source/Evolve/WarpXEvolve.cpp]: This needs to be inside the if block, because the Electromagnetic case has already done the Esolve at this point (it's in `OneStep_nosub`, so we only want this for electrostatic.

File: Source/Evolve/WarpXEvolve.cpp [hunk 2]
```diff
@@ -249,7 +250,9 @@ WarpX::Evolve (int numsteps)
             // positions are next pushed) so that the particles do not deposit out of bounds
             // and so that the fields are at the correct time in the output.
             bool const reset_fields = true;
-            ComputeSpaceChargeField( reset_fields );
+            if (warpx_py_fieldsolver) warpx_py_fieldsolver();
+                else AddSpaceChargeFieldLabFrame();
```
Comment (peterscherpelz) [inline Source/Evolve/WarpXEvolve.cpp]: Indentation needs to be fixed - neither the else nor the line below should be further indented.

**Review by peterscherpelz — CHANGES_REQUESTED** [review]
I would check before the whole per-step loop in Evolve - like line 41ish - that we didn't install fieldsolver when we're not using an electrostatic solver, as that would be a combination that's very non-intuitive for the user (the fieldsolver doesn't work). eg (wrong syntax, but the idea):
if ( do_electrastic == Ele...Algo::None and fieldsolver_is_installed):
    amrex::Abort("Bad combination!")

We'll need to put more of the embedded boundaries splits @roelof-groenewald made for this to make sense, because this version of the code can't disable the built-in electrostatic solver, which we need to do if a python solver is installed.

I would recommend splitting the beforeEsolve callback insertion into an upstream PR, and we'll hold this PR to include the other EB code that disables built-in-solver too.

File: Source/FieldSolver/ElectrostaticSolver_EB.cpp [hunk 3]
```diff
@@ -0,0 +1,658 @@
+/* Copyright 2021 Modern Electron
```
Comment (peterscherpelz) [inline Source/FieldSolver/ElectrostaticSolver_EB.cpp]: Can we move this file to ElectrostaticSolver_ME ? Then it's consistent with the compile var, and we won't get confused when EB is introduced upstream.

File: Source/FieldSolver/ElectrostaticSolver_EB.cpp [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+                },
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    Bz_arr(i,j,k) += inv_c * (
+                        +beta_y*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) ));
+                }
+            );
+#endif
+        }
+    }
+}
```
Comment (peterscherpelz) [inline Source/FieldSolver/ElectrostaticSolver_EB.cpp]: I believe newline at end of file is standard.

File: Source/FieldSolver/Make.package [hunk 5]
```diff
@@ -1,5 +1,9 @@
 CEXE_sources += WarpXPushFieldsEM.cpp
-CEXE_sources += ElectrostaticSolver.cpp
+ifeq ($(ME_Solver),TRUE)
+  CEXE_sources += ElectrostaticSolver_EB.cpp
```
Comment (peterscherpelz) [inline Source/FieldSolver/Make.package]: Change name to <same>_ME.cpp here too.

File: mewarpx/examples/test_install_fieldsolver.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+#     plt.plot(np.mean(phi_data, axis=1), 'o-')
+#     plt.plot(np.mean(phi, axis=1), 'o-')
+
+#     #plt.plot(phi_data[:,4], 'o-')
+#     #plt.plot(phi[:,4], 'o-')
+#     plt.show()
+
+# callbacks.installfieldsolver(dummy_solve)
+
+sim.step(5)
```
Comment (peterscherpelz) [inline mewarpx/examples/test_install_fieldsolver.py]: newline here too

File: mewarpx/examples/test_install_fieldsolver.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    particle_type='electron', name='electrons',
+    initial_distribution=uniform_plasma_elec
+)
+ions = picmi.Species(
+    particle_type='He', name='he_ions',
+    charge='q_e',
+    initial_distribution=uniform_plasma_ion
+)
+
+''' Add back when MCC initialization has been added to mewarpx
```
Comment (peterscherpelz) [inline mewarpx/examples/test_install_fieldsolver.py]: For this sample that just tests installfieldsolver, eliminate this commented section - and the extraneous stuff at bottom of this script

File: Python/pywarpx/callbacks.py [hunk 8]
```diff
@@ -28,6 +28,8 @@
  - afterdeposition <installafterdeposition>: after particle deposition (for charge and/or current)
  - beforestep <installbeforestep>: before the time step
  - afterstep <installafterstep>: after the time step
+ - fieldsolver <installfieldsolver>: In place of the AddSpaceChargeFieldLabFrame() call but only in
+                                     an electrostatic simulation with WarpX_EB = ON
```
Comment (roelof-groenewald) [inline Python/pywarpx/callbacks.py]: The flag is no longer WarpX_EB.

File: Source/Evolve/WarpXEvolve.cpp [hunk 9]
```diff
@@ -268,15 +268,30 @@ WarpX::Evolve (int numsteps)
...
+
         if( do_electrostatic != ElectrostaticSolverAlgo::None ) {
             // Electrostatic solver:
             // For each species: deposit charge and add the associated space-charge
             // E and B field to the grid ; this is done at the end of the PIC
             // loop (i.e. immediately after a `Redistribute` and before particle
             // positions are next pushed) so that the particles do not deposit out of bounds
             // and so that the fields are at the correct time in the output.
-            bool const reset_fields = true;
+            const bool reset_fields = true;
```
Comment (roelof-groenewald) [inline Source/Evolve/WarpXEvolve.cpp]: Why this change?

File: Source/Evolve/WarpXEvolve.cpp [hunk 10]
```diff
@@ -268,15 +268,30 @@ WarpX::Evolve (int numsteps)
...
+
+#ifdef ME_SOLVER
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                do_electrostatic == ElectrostaticSolverAlgo::LabFrame,
+                "Embedded boundary solver can only be used in the labframe."
```
Comment (roelof-groenewald) [inline Source/Evolve/WarpXEvolve.cpp]:
```suggestion
                "ME solver can only be used in the labframe."
```

File: Source/Make.WarpX [hunk 11]
```diff
@@ -99,6 +99,15 @@ ifeq ($(USE_EB),TRUE)
...
 
+ifeq ($(ME_SOLVER),TRUE)
+  include $(WARPX_HOME)/Source/FieldSolver/Make.package
+  CXXFLAGS += -DME_SOLVER
+  CFLAGS   += -DME_SOLVER
+  FFLAGS   += -DME_SOLVER
+  F90FLAGS += -DME_SOLVER
+  USERSuffix := $(USERSuffix).EB
+endif
+
```
Comment (roelof-groenewald) [inline Source/Make.WarpX]: Why do we need this here? The FieldSolver/Make.package file should always be included. Do we actually need the `DME_SOLVER` flags for all those cases? I'm pretty sure the `USERSuffix := $(USERSuffix).EB` line is not needed.

**Review by roelof-groenewald — CHANGES_REQUESTED** [review]
Left a couple of comments of some initial questions.

**Comment by KZhu-ME** [comment]
All the failed azure tests failed when trying to run `python setup.py install --with-libwarpx 2d` in the `Python` directory. This command starts with an error saying that the repository is not a git repository:
```
fatal: not a git repository (or any of the parent directories): .git
```
After giving a lot of warnings, it ends with the error message saying that it can't create or remove files in the install directory:
```
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 13] Permission denied: '/usr/local/lib/python3.6/dist-packages/test-easy-install-17179.write-test'

The installation directory you s pecified (via --install-dir, --prefix, or
the distutils default setting) was:

    /usr/local/lib/python3.6/dist-packages/

Perhaps your account does not have write access to this directory?  If the
installation directory is a system-owned directory, you may need to sign in
as the administrator or "root" account.  If you do not have administrative
access to this machine, you may wish to choose a different installation
directory, preferably one that is listed in your PYTHONPATH environment
variable.

For information on other options, you may wish to consult the
documentation at:

  https://setuptools.readthedocs.io/en/latest/easy_install.html

Please make the appropriate changes for your system and try again.

make: *** [installwarpx] Error 1
```

**Comment by roelof-groenewald** [comment]
@kzhu-ME the Azure test builds still uses the GNUmakefile so we should avoid making any changes to it. The tests failed here (except the Python ones) since the `USE_PYTHON_MAIN` flag was changed to True.

File: Python/pywarpx/callbacks.py [hunk 12]
```diff
@@ -372,6 +378,24 @@ def isinstalledafterdeposition(f):
     "Checks if the function is called after a particle deposition"
     return _afterdeposition.isinstalledfuncinlist(f)
 
+# ----------------------------------------------------------------------------
+def callfromfieldsolver(f):
+    installfieldsolver(f)
+    return f
+def installfieldsolver(f):
+    """Adds a Adds a function to perform the field solve. Note that the C++ object
```
Comment (roelof-groenewald) [inline Python/pywarpx/callbacks.py]:
```suggestion
    """Adds a function to perform the field solve. Note that the C++ object
```

File: Source/Evolve/WarpXEvolve.cpp [hunk 13]
```diff
@@ -268,6 +268,7 @@ WarpX::Evolve (int numsteps)
             mypc->SortParticlesByBin(sort_bin_size);
         }
 
+
```
Comment (roelof-groenewald) [inline Source/Evolve/WarpXEvolve.cpp]:
```suggestion
```

**Review by roelof-groenewald — APPROVED** [review]
I left suggestions for quick changes. Otherwise it looks good to me but please remember to address @peterscherpelz's comments about the test input file.

File: Source/FieldSolver/ElectrostaticSolver_ME.cpp [hunk 14]
```diff
@@ -0,0 +1,658 @@
...
+                   amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                   std::array<Real, 3> const beta,
+                   Real const required_precision,
+                   int const max_iters) const
+{
+#ifdef WARPX_DIM_RZ
+    computePhiRZ( rho, phi, beta, required_precision, max_iters );
+#else
+    computePhiCartesian( rho, phi, beta, required_precision, max_iters );
+#endif
```
Comment (roelof-groenewald) [inline Source/FieldSolver/ElectrostaticSolver_ME.cpp:115]: Please also incorporate @MKieburtz's changes here that allows us to set the field solve verbosity.

**Review by EtherealMC-Bit — COMMENTED** [review]
Sorry for the confusion Kevin but this was the wrong way I went about doing the verbosity first. Here's the right way in this commit and pr. https://github.com/ECP-WarpX/WarpX/pull/2041/commits/31394ab2fdf0eadcec24041e4e30ef71cd308f96

**Comment by EtherealMC-Bit** [comment]
> Sorry for the confusion Kevin but this was the wrong way I went about doing the verbosity first. Here's the right way in this commit and pr. [ECP-WarpX@31394ab](https://github.com/ECP-WarpX/WarpX/commit/31394ab2fdf0eadcec24041e4e30ef71cd308f96)

Never mind

**Comment by nruof** [comment]
After the verbosity changes that should be good

**Comment by KZhu-ME** [comment]
Is this branch ok to merge now?

**Review by peterscherpelz — APPROVED** [review]

**Comment by peterscherpelz** [comment]
With two approvals and @MKieburtz 's existing review I'll go ahead and merge. Since there were a lot of commits, I'm going to squash them to keep our history cleaner. We can discuss if we want to make that normal practice, or keep all commits in PRs.
